# Credit Card Fraud Detection

In [5]:
# Importing relevant libraries
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

In [6]:
# Loading the dataset
credit_card_data = pd.read_csv('creditcard.csv')

In [7]:
credit_card_data.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [8]:
credit_card_data.tail()

Time         V1         V2        V3        V4        V5  \
284802  172786.0 -11.881118  10.071785 -9.834783 -2.066656 -5.364473   
284803  172787.0  -0.732789  -0.055080  2.035030 -0.738589  0.868229   
284804  172788.0   1.919565  -0.301254 -3.249640 -0.557828  2.630515   
284805  172788.0  -0.240440   0.530483  0.702510  0.689799 -0.377961   
284806  172792.0  -0.533413  -0.189733  0.703337 -0.506271 -0.012546   

              V6        V7        V8        V9  ...       V21       V22  \
284802 -2.606837 -4.918215  7.305334  1.914428  ...  0.213454  0.111864   
284803  1.058415  0.024330  0.294869  0.584800  ...  0.214205  0.924384   
284804  3.031260 -0.296827  0.708417  0.432454  ...  0.232045  0.578229   
284805  0.623708 -0.686180  0.679145  0.392087  ...  0.265245  0.800049   
284806 -0.649617  1.577006 -0.414650  0.486180  ...  0.261057  0.643078   

             V23       V24       V25       V26       V27       V28  Amount  \
284802  1.014480 -0.509348  1.436807  0.250034  0.943651  0.823731    0.77   
284803  0.012463 -1.016226 -0.606624 -0.395255  0.068472 -0.053527   24.79   
284804 -0.037501  0.640134  0.265745 -0.087371  0.004455 -0.026561   67.88   
284805 -0.163298  0.123205 -0.569159  0.546668  0.108821  0.104533   10.00   
284806  0.376777  0.008797 -0.473649 -0.818267 -0.002415  0.013649  217.00   

        Class  
284802      0  
284803      0  
284804      0  
284805      0  
284806      0  

[5 rows x 31 columns]

In [10]:
credit_card_data.shape

(284807, 31)

In [9]:
credit_card_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

In [11]:
credit_card_data.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

- That means there are no missing values in the dataset

In [12]:
credit_card_data['Class'].value_counts()

0    284315
1       492
Name: Class, dtype: int64

- There are 284315 normal transactions and 492 fraudulent transactions
- Dataset is highly imbalanced.

In [13]:
# Separating the data for analysis
legit = credit_card_data[credit_card_data.Class == 0]
fraud = credit_card_data[credit_card_data.Class == 1]

In [14]:
legit.shape

(284315, 31)

In [15]:
fraud.shape

(492, 31)

In [19]:
legit.Amount.describe()

count    284315.000000
mean         88.291022
std         250.105092
min           0.000000
25%           5.650000
50%          22.000000
75%          77.050000
max       25691.160000
Name: Amount, dtype: float64

In [20]:
fraud.Amount.describe()

count     492.000000
mean      122.211321
std       256.683288
min         0.000000
25%         1.000000
50%         9.250000
75%       105.890000
max      2125.870000
Name: Amount, dtype: float64

- We can see that fraudulent transactions mean amount are higher than legit transcations but maximum amount compared to legit transcations is really low.

In [21]:
credit_card_data.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      94838.202258  0.008258 -0.006271  0.012171 -0.007860  0.005453   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0      0.002419  0.009637 -0.000987  0.004467  ... -0.000644 -0.001235   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0     -0.000024  0.000070  0.000182 -0.000072 -0.000089 -0.000295 -0.000131   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       88.291022  
1      122.211321  

[2 rows x 30 columns]

- The difference in the mean values between groups will be reall beneficial in the machine learning part.

## Under-Sampling

- Building a sample dataset containing similar distribution of legit transactions and fraudulent transactions
- Number of fraudulent transactions --> 492

In [22]:
# taking random 492 points from the legit variable that contains similar distribution
legit_sample = legit.sample(n = 492)

In [23]:
# concatenating two dataframes
new_dataset = pd.concat([legit_sample, fraud], axis = 0)

In [24]:
new_dataset.head()

Time        V1        V2        V3        V4        V5        V6  \
62952    50488.0  1.236678  0.401743  0.103536  1.124861 -0.131004 -1.065329   
237161  149124.0 -0.448410  0.974965  0.249840 -1.080694  1.118870 -0.493610   
72214    54615.0 -5.195481  6.436327 -4.477853  0.083120 -0.345979 -2.095891   
64734    51283.0  1.050876  0.217431  1.441137  2.359725 -0.558371  0.683278   
255667  157338.0  0.087776  0.972786 -0.335244 -0.623527  0.885457 -0.585316   

              V7        V8        V9  ...       V21       V22       V23  \
62952   0.400262 -0.294452 -0.148828  ... -0.022416 -0.038405 -0.103692   
237161  1.100816  0.049111 -0.555809  ... -0.219332 -0.728306 -0.051170   
72214   0.607937  0.151704  4.471110  ... -0.850490 -0.597099  0.447240   
64734  -0.720793  0.411017 -0.429787  ...  0.157033  0.413962  0.054542   
255667  0.882137  0.031666 -0.303650  ... -0.280207 -0.706277  0.109367   

             V24       V25       V26       V27       V28  Amount  Class  
62952   0.397283  0.727783 -0.359738  0.002095  0.018780   14.32      0  
237161  0.587857 -0.337811  0.032872  0.052939  0.155409    9.85      0  
72214   0.076942  0.878953 -0.311894  3.067438  1.641677    0.89      0  
64734  -0.011893  0.112929  0.003944  0.043541  0.024707    4.95      0  
255667  0.657706 -0.470047  0.086840  0.216750  0.075599    6.99      0  

[5 rows x 31 columns]

In [25]:
new_dataset['Class'].value_counts()

0    492
1    492
Name: Class, dtype: int64

In [21]:
credit_card_data.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      94838.202258  0.008258 -0.006271  0.012171 -0.007860  0.005453   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0      0.002419  0.009637 -0.000987  0.004467  ... -0.000644 -0.001235   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0     -0.000024  0.000070  0.000182 -0.000072 -0.000089 -0.000295 -0.000131   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       88.291022  
1      122.211321  

[2 rows x 30 columns]

In [26]:
new_dataset.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      97834.810976 -0.104377 -0.004870  0.058890 -0.099035  0.053070   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0      0.017510 -0.024673  0.003257  0.049540  ... -0.029138 -0.000203   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0      0.010000  0.051263 -0.002633  0.003815 -0.013647  0.023974 -0.004875   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       83.328740  
1      122.211321  

[2 rows x 30 columns]

- Our new dataset contains similar distribution as shown above, '.sample' function works like a charm.

## Splitting the data into Features and Targets

In [27]:
X = new_dataset.drop(columns = 'Class', axis = 1)
Y = new_dataset['Class']

In [28]:
X

Time        V1        V2        V3        V4        V5        V6  \
62952    50488.0  1.236678  0.401743  0.103536  1.124861 -0.131004 -1.065329   
237161  149124.0 -0.448410  0.974965  0.249840 -1.080694  1.118870 -0.493610   
72214    54615.0 -5.195481  6.436327 -4.477853  0.083120 -0.345979 -2.095891   
64734    51283.0  1.050876  0.217431  1.441137  2.359725 -0.558371  0.683278   
255667  157338.0  0.087776  0.972786 -0.335244 -0.623527  0.885457 -0.585316   
...          ...       ...       ...       ...       ...       ...       ...   
279863  169142.0 -1.927883  1.125653 -4.518331  1.749293 -1.566487 -2.010494   
280143  169347.0  1.378559  1.289381 -5.004247  1.411850  0.442581 -1.326536   
280149  169351.0 -0.676143  1.126366 -2.213700  0.468308 -1.120541 -0.003346   
281144  169966.0 -3.113832  0.585864 -5.399730  1.817092 -0.840618 -2.943548   
281674  170348.0  1.991976  0.158476 -2.583441  0.408670  1.151147 -0.096695   

              V7        V8        V9  ...       V20       V21       V22  \
62952   0.400262 -0.294452 -0.148828  ... -0.123089 -0.022416 -0.038405   
237161  1.100816  0.049111 -0.555809  ... -0.156084 -0.219332 -0.728306   
72214   0.607937  0.151704  4.471110  ...  3.676496 -0.850490 -0.597099   
64734  -0.720793  0.411017 -0.429787  ... -0.177015  0.157033  0.413962   
255667  0.882137  0.031666 -0.303650  ...  0.028200 -0.280207 -0.706277   
...          ...       ...       ...  ...       ...       ...       ...   
279863 -0.882850  0.697211 -2.064945  ...  1.252967  0.778584 -0.319189   
280143 -1.413170  0.248525 -1.127396  ...  0.226138  0.370612  0.028234   
280149 -2.234739  1.210158 -0.652250  ...  0.247968  0.751826  0.834108   
281144 -2.208002  1.058733 -1.632333  ...  0.306271  0.583276 -0.269209   
281674  0.223050 -0.068384  0.577829  ... -0.017652 -0.164350 -0.295135   

             V23       V24       V25       V26       V27       V28  Amount  
62952  -0.103692  0.397283  0.727783 -0.359738  0.002095  0.018780   14.32  
237161 -0.051170  0.587857 -0.337811  0.032872  0.052939  0.155409    9.85  
72214   0.447240  0.076942  0.878953 -0.311894  3.067438  1.641677    0.89  
64734   0.054542 -0.011893  0.112929  0.003944  0.043541  0.024707    4.95  
255667  0.109367  0.657706 -0.470047  0.086840  0.216750  0.075599    6.99  
...          ...       ...       ...       ...       ...       ...     ...  
279863  0.639419 -0.294885  0.537503  0.788395  0.292680  0.147968  390.00  
280143 -0.145640 -0.081049  0.521875  0.739467  0.389152  0.186637    0.76  
280149  0.190944  0.032070 -0.739695  0.471111  0.385107  0.194361   77.89  
281144 -0.456108 -0.183659 -0.328168  0.606116  0.884876 -0.253700  245.00  
281674 -0.072173 -0.450261  0.313267 -0.289617  0.002988 -0.015309   42.53  

[984 rows x 30 columns]

In [29]:
Y

62952     0
237161    0
72214     0
64734     0
255667    0
         ..
279863    1
280143    1
280149    1
281144    1
281674    1
Name: Class, Length: 984, dtype: int64

In [30]:
# Splitting data into training and testing data
X_train, X_test, Y_train, Y_test = train_test_split(X,Y, test_size = 0.2, stratify = Y, random_state = 42)

In [31]:
print(X.shape, X_train.shape, X_test.shape)

(984, 30) (787, 30) (197, 30)


## Model Training

In [32]:
model = LogisticRegression()

In [33]:
model.fit(X_train,Y_train)

LogisticRegression()

## Model Evaluation

In [34]:
# accuracy on training data
X_train_pred = model.predict(X_train)
training_data_accuracy = accuracy_score(X_train_pred, Y_train)

In [35]:
print('Accuracy on Training data: ', training_data_accuracy)

Accuracy on Training data:  0.9440914866581956


In [36]:
# accuracy on testing data
X_test_pred = model.predict(X_test)
test_data_accuracy = accuracy_score(X_test_pred, Y_test)

In [37]:
print('Accuracy on testing data: ', test_data_accuracy)

Accuracy on testing data:  0.9593908629441624


- The model works perfectly well.